ed448-goldilocks/src/montgomery/point.rs

```diff
@@ -0,0 +1,580 @@
+    }
+
+    fn identity() -> Self {
+        Self::IDENTITY
+    }
+
+    fn generator() -> Self {
+        Self::GENERATOR
+    }
+
+    fn is_identity(&self) -> Choice {
+        self.ct_eq(&Self::IDENTITY)
+    }
+
+    // See Complete Addition Law for Montgomery Curves - Algorithm 3.
+    // Slightly corrected from the derivation in the same paper.
+    fn double(&self) -> Self {
+        const A_MINUS_1: FieldElement = FieldElement(ConstMontyType::new(&U448::from_u64(156325)));
+
+        let (x, y, z) = (self.U, self.V, self.W);
+
+        let t0 = x.square();
+        let t1 = y.square();
+        let t2 = z.square();
+        let t3 = (x + y).square();
+        let t4 = (y + z).square();
+        let t5 = (x + z).square();
+        let t6 = t1 + t2;
+        let t7 = (t0 - t2).double();
+        let t8 = A_MINUS_1 * t0;
+        let t9 = t0 - t1;
+        let t10 = FieldElement::J * (t5 - t2) + t0 + t9;
+        let t11 = t5 + t8;
+        let t13 = t6.double(); // corrected - replaces t12
+
+        let S_MINUS_U = t3 - t6;
+        let S_PLUS_U = -S_MINUS_U + t7;
+        let R_MINUS_T = t4 - t11;
+        let R_PLUS_T = t4 - t13 + t11; // corrected
+        let W_MINUS_S = t11 - t9;
+        let W_PLUS_S = W_MINUS_S + t7;
+        let T_MINUS_V = t11 - t10 - t13 + t8; // corrected
+        let T_PLUS_V = t5 + t10;
+        let U_MINUS_W = S_PLUS_U - W_PLUS_S;
+        let R_PLUS_V = R_MINUS_T + T_PLUS_V;
+
+        let C = R_PLUS_T * S_MINUS_U;
+        let D = R_MINUS_T * S_PLUS_U;
+        let E = T_PLUS_V * W_MINUS_S;
+        let F = T_MINUS_V * W_PLUS_S;
+        let X = C + D;
+        let Y = E + F;
+        let Z = U_MINUS_W.double() * R_PLUS_V + C - D + E - F;
+
+        Self { U: X, V: Y, W: Z }
+    }
+}
+
+impl CurveGroup for ProjectiveMontgomeryPoint {
+    type AffineRepr = AffineMontgomeryPoint;
+
+    fn to_affine(&self) -> Self::AffineRepr {
+        let W_inv = self.W.invert();
+        let U = self.U * W_inv;
+        let V = self.V * W_inv;
+
+        AffineMontgomeryPoint { U, V }
+    }
+}
+
+impl GroupEncoding for ProjectiveMontgomeryPoint {
+    type Repr = Array<u8, U57>;
+
+    fn from_bytes(bytes: &Self::Repr) -> CtOption<Self> {
+        // Safe to unwrap here as the underlying data structure is an array
+        let (tag, bytes) = bytes.split_first().expect("slice is non-empty");
+
+        let mut x_bytes: [u8; 56] = [0; 56];
+        x_bytes.copy_from_slice(bytes);
+
+        let (sign, valid) = match *tag {
+            0x02 => (Choice::from(0), Choice::from(1)),
+            0x03 => (Choice::from(1), Choice::from(1)),
+            _ => (Choice::from(0), Choice::from(0)),
+        };
+
+        FieldElement::from_repr(&x_bytes).and_then(|U| {
+            CtOption::new(
+                ProjectiveMontgomeryXpoint {
+                    U,
+                    W: FieldElement::ONE,
+                }
+                .to_extended(sign),
+                valid,
+            )
+        })
+    }
```

Comment:
Just wanted to highlight that I'm using SEC1 encoding here considering that I found no standard at all to encode full-coordinate Curve448.

Alternatively we can just follow Edwards448 here.

Reply:
These curves definitely don't use SEC1. You can shove the sign bit in the most significant bit (i.e. bit 255), but that's nonstandard as it's specified to be set to 0 (at least for X25519).

Reply:
Yeah that's why I went with SEC1, because the SEC1 specification doesn't say anywhere its limited to Weierstrass curves.

But I'm happy to follow the route you are suggesting here.

Reply:
Done.